### ISSUE TITLE
[remote_wal] Add transaction support for producing records

### ISSUE THREAD

**Issue by niebayes** [GreptimeTeam/greptimedb#3108]
### What type of enhancement is this?

Tech debt reduction

### What does the enhancement do?

For records generated by splitting a large entry, we need to ensure they're produced jointly. If producing one of them fails, we need to abort and properly handle records that were already produced.

**Comment by niebayes** [comment]
For an excessively large log entry, the write-ahead log (WAL) based on Kafka splits it into multiple records during the write process, subsequently writing them individually. In this process, certain records may fail to be written, while others succeed. Since Kafka provides an append-only semantics, there will be some record leftover in Kafka. These leftover will impact record consumption. To address this problem, we plan to perform the following operations:

- During Production:
  - When producing records obtained from splitting a large log entry, we set their timestamps to the same value, indicating that they belong to the same log entry.
  - If the production of any single record fails, we consider the entire batch of records as failed. The `append_batch` simply returns a producing error without taking any specific actions on Kafka. For example, we will not append any undo logs to Kafka.
  
- During Consumption:
  - Previously, we aggregated records belonging to the same log entry based on the entry id. However, since a production failure might result in the repetition of a record belong to the same entry id in Kafka, we now determine the most recent records based on their timestamps. Specifically, newer records will replace the older ones.
  - Previously, after consuming a batch of records, we check for incomplete record sequences, and if found, a consume error is returned. Now, we only print a warning without throwing an error. This is because, in the case of production failures, we lack the capability to clean up incomplete records in Kafka.
